File: go/parser/kind7375.go
// kind7375.go - Token Event

package parser

import (
	"encoding/json"
	"fmt"

	"github.com/candypoets/nutscash/types"
	"github.com/nbd-wtf/go-nostr"
)

// Kind7375Parsed represents a parsed token event
type Kind7375Parsed struct {
	MintURL    string             `json:"mintUrl" msgpack:"mintUrl"`
	Proofs     []types.ProofUnion `json:"proofs" msgpack:"proofs"`
	DeletedIDs []string           `json:"deletedIds,omitempty" msgpack:"deletedIds,omitempty"` // IDs of token events that were deleted
	Decrypted  bool               `json:"decrypted" msgpack:"decrypted"`                       // Whether the content was successfully decrypted
}

// TokenContent represents the encrypted content of a token event
type TokenContent struct {
	Mint   string             `json:"mint" msgpack:"mint"`
	Id     string             `json:"id" msgpack:"id"`
	Proofs []types.ProofUnion `json:"proofs" msgpack:"proofs"`
	Del    []string           `json:"del,omitempty" msgpack:"del,omitempty"`
}

// ParseKind7375 parses a kind 7375 (token) event
func (p *Parser) ParseKind7375(event nostr.Event) (*Kind7375Parsed, *[]types.Request, error) {
	if event.Kind != 7375 {
		return nil, nil, fmt.Errorf("event is not kind 7375")
	}

	parsed := &Kind7375Parsed{
		Decrypted: false,
	}

	// Try to decrypt the content if signer is available
	if p.Signer != nil {
		pubkey, _ := p.Signer.GetPublicKey()
		decrypted, err := p.Signer.NIP44Decrypt(pubkey, event.Content)
		if err == nil && decrypted != "" {
			var content TokenContent
			if err := json.Unmarshal([]byte(decrypted), &content); err == nil {
				parsed.MintURL = content.Mint
				parsed.Proofs = content.Proofs
				parsed.DeletedIDs = content.Del
				parsed.Decrypted = true
			}
		}
	}

	return parsed, nil, nil
}

// PrepareKind7375 prepares a kind 7375 event for publishing by encrypting content and signing
func (p *Parser) PrepareKind7375(event *nostr.Event) error {
	if event.Kind != 7375 {
		return fmt.Errorf("event is not kind 7375")
	}

	// Content must be a valid JSON for TokenContent
	var content TokenContent
	if err := json.Unmarshal([]byte(event.Content), &content); err != nil {
		return fmt.Errorf("invalid token content: %w", err)
	}

	// Validate content
	if content.Mint == "" {
		return fmt.Errorf("token content must specify a mint")
	}

	if len(content.Proofs) == 0 {
		return fmt.Errorf("token content must include at least one proof")
	}

	// NIP-44 encrypt the content
	contentJSON, err := json.Marshal(content)
	if err != nil {
		return fmt.Errorf("failed to marshal token content: %w", err)
	}
	pubkey, _ := p.Signer.GetPublicKey()

	encrypted, err := p.Signer.NIP44Encrypt(pubkey, string(contentJSON))
	if err != nil {
		return fmt.Errorf("failed to encrypt token content: %w", err)
	}
	event.Content = encrypted

	// Sign the event
	return p.Signer.SignEvent(event)
}

File: src/model/nostr-worker/src/parser/kind7375.rs
use crate::parser::Parser;
use crate::types::network::Request;
use anyhow::{anyhow, Result};
use nostr::{Event, UnsignedEvent};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tracing::{info, warn};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DLEQProof {
    pub e: String,
    pub s: String,
    pub r: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProofUnion {
    pub amount: u64,
    pub id: String,
    pub secret: String,
    #[serde(rename = "C")]
    pub c: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dleq: Option<DLEQProof>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub version: Option<u8>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Kind7375Parsed {
    #[serde(rename = "mintUrl")]
    pub mint_url: String,
    pub proofs: Vec<ProofUnion>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(rename = "deletedIds")]
    pub deleted_ids: Vec<String>,
    pub decrypted: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct TokenContent {
    pub mint: String,
    pub id: String,
    pub proofs: Vec<ProofUnion>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub del: Option<Vec<String>>,
}

impl Parser {
    pub fn parse_kind_7375(&self, event: &Event) -> Result<(Kind7375Parsed, Option<Vec<Request>>)> {
        if event.kind.as_u64() != 7375 {
            return Err(anyhow!("event is not kind 7375"));
        }

        let mut parsed = Kind7375Parsed {
            mint_url: String::new(),
            proofs: Vec::new(),
            deleted_ids: Vec::new(),
            decrypted: false,
        };

        let signer = &self.signer_manager;

        if signer.has_signer() {
            let pubkey = signer.get_public_key()?;
            if let Ok(decrypted) = signer.nip44_decrypt(&pubkey, &event.content) {
                if !decrypted.is_empty() {
                    if let Ok(content) = serde_json::from_str::<TokenContent>(&decrypted) {
                        parsed.mint_url = content.mint;
                        parsed.proofs = content.proofs;
                        parsed.deleted_ids = content.del.unwrap_or_default();
                        parsed.decrypted = true;
                    } else if let Err(e) = serde_json::from_str::<TokenContent>(&decrypted) {
                        warn!("Failed to parse 7375 token content: {}", e);
                    }
                }
            }
        } else {
            warn!("No signer found for event 7375");
        }

        Ok((parsed, None))
    }

    pub fn prepare_kind_7375(&self, unsigned_event: &mut UnsignedEvent) -> Result<Event> {
        if unsigned_event.kind.as_u64() != 7375 {
            return Err(anyhow!("event is not kind 7375"));
        }

        // Content must be a valid JSON for TokenContent
        let _content: TokenContent = serde_json::from_str(&unsigned_event.content)
            .map_err(|e| anyhow!("invalid token content: {}", e))?;

        // Validate content
        if _content.mint.is_empty() {
            return Err(anyhow!("token content must specify a mint"));
        }

        if _content.proofs.is_empty() {
            return Err(anyhow!("token content must include at least one proof"));
        }

        let signer = &self.signer_manager;

        if !signer.has_signer() {
            return Err(anyhow!("no signer available for encryption"));
        }

        let pubkey = signer.get_public_key()?;
        let encrypted_content = signer.nip44_encrypt(&pubkey, &unsigned_event.content)?;

        unsigned_event.content = encrypted_content;

        let event = signer.sign_event(unsigned_event)?;

        Ok(event)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use nostr::{EventBuilder, Keys, Kind};

    #[test]
    fn test_parse_kind_7375_basic() {
        let keys = Keys::generate();

        let event = EventBuilder::new(Kind::Custom(7375), "encrypted_content", Vec::new())
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let (parsed, _) = parser.parse_kind_7375(&event).unwrap();

        assert!(!parsed.decrypted); // No decryption without signer
        assert!(parsed.mint_url.is_empty());
        assert!(parsed.proofs.is_empty());
    }

    #[test]
    fn test_prepare_kind_7375_invalid_content() {
        let keys = Keys::generate();

        let mut event = EventBuilder::new(Kind::Custom(7375), "invalid json", Vec::new())
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let result = parser.prepare_kind_7375(&mut event);

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("invalid token content"));
    }

    #[test]
    fn test_parse_wrong_kind() {
        let keys = Keys::generate();

        let event = EventBuilder::new(Kind::TextNote, "test", Vec::new())
            .to_event(&keys)
            .unwrap();

        let parser = Parser::default();
        let result = parser.parse_kind_7375(&event);

        assert!(result.is_err());
    }
}
